include/fastcdr/xcdr/optional.hpp

```diff
@@ -0,0 +1,400 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#ifndef _FASTCDR_XCDR_OPTIONAL_HPP_
+#define _FASTCDR_XCDR_OPTIONAL_HPP_
+
+#include <new>
+#include <type_traits>
+
+#include "../exceptions/BadOptionalAccessException.hpp"
+
+namespace eprosima {
+namespace fastcdr {
+
+namespace detail {
+template<class T, class E = void>
+struct optional_storage
+{
+    union
+    {
+        char dummy_; T val_;
+    };
+
+    bool engaged_ { false };
+
+    optional_storage()
+    {
+    }
+
+    ~optional_storage()
+    {
+        if (engaged_)
+        {
+            val_.~T();
+        }
+    }
+
+};
+
+/* *INDENT-OFF* */
+template<class T>
+struct optional_storage<T, typename std::enable_if<std::is_trivially_destructible<T>{}>::type>
+{
+    union
+    {
+        char dummy_; T val_;
+    };
+
+    bool engaged_ { false };
+
+    optional_storage()
+    {
+    }
+
+    ~optional_storage() = default;
+};
+/* *INDENT-ON* */
+} // namespace detail
+
+//! An empty class type used to indicate optional type with uninitialized state.
+struct nullopt_t
+{
+    constexpr explicit nullopt_t(
+            int)
+    {
+    }
+
+};
+
+/*!
+ * @brief nullopt is a constant of type nullopt_t that is used to indicate optional type with uninitialized state.
+ */
+static constexpr nullopt_t nullopt {0};
+
+/*!
+ * @brief This class template manages an optional contained value, i.e. a value that may or may not be present.
+ */
+template<class T>
+class optional
+{
+public:
+
+    using type = T;
+
+    //! Default constructor
+    optional() = default;
+
+    //! Copy constructor from an instance of the templated class.
+    optional(
+            const T& val) noexcept
+    {
+        ::new(&storage_.val_)T(val);
+        storage_.engaged_ = true;
+    }
+
+    //! Move constructor from an instance of the templated class.
+    optional(
+            T&& val) noexcept
+    {
+        ::new(&storage_.val_)T(std::move(val));
+        storage_.engaged_ = true;
+    }
+
+    //! Copy constructor.
+    optional(
+            const optional<T>& val) noexcept
+    {
+        ::new(&storage_.val_)T(val.storage_.val_);
+        storage_.engaged_ = val.storage_.engaged_;
+    }
+
+    //! Move constructor.
+    optional(
+            optional<T>&& val) noexcept
+    {
+        ::new(&storage_.val_)T(std::move(val.storage_.val_));
+        storage_.engaged_ = val.storage_.engaged_;
+    }
+
+    //! Destructor
+    ~optional() = default;
+
+    /*!
+     * @brief Constructs the contained value in-place
+     *
+     * @param[in] _args The arguments to pass to the constructor.
+     */
+    template<class ... Args> void emplace(
+            Args&&... _args)
+    {
+        reset();
+        storage_.val_.T(std::forward<Args>(_args)...);
```

Comment:
`std::forward` is defined in header `<utility>` which has not been included.

Reply:
Done in e3a6645